Improve repeat_interleave with scalar repeat value

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102570

`repeat_interleave_symint` is currently implemented by guarding on the `SymInt`
and converting it to a tensor to pass to the Tensor overload. This instead
implements it as a copy of an expanded tensor, which can be done without guards
and is also much more efficient in eager mode to boot.

For example, these are timings for `x.repeat_interleave(100, dim=-1)` with `x.shape == (1000, 100)`

| Device | Time (Master) | Time (This PR)  | Speedup |
|--------|---------------|-----------------|---------|
| cpu    | 18.8 ms       | 3.5 ms          | 5.4     |
| cuda   | 271 us        | 134 us          | 2.0     |

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

peterbell10 committed:
Improve repeat_interleave with scalar repeat value

`repeat_interleave_symint` is currently implemented by guarding on the `SymInt`
and converting it to a tensor to pass to the Tensor overload. This instead
implements it as a copy of an expanded tensor, which can be done without guards
and is also much more efficient in eager mode to boot.

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/Repeat.cpp:
```diff
@@ -97,31 +97,30 @@ Tensor repeat_interleave(
+                calculated_size, " but got ", *output_size);
   }
 
+  return input.clone().flatten(dim, dim + 1);
```
This is potentially copying twice. What about doing the flatten first, then check if we copied, and if we didn't then clone?

peterbell10 replied:
I've switched it to force contiguous format, which should guarantee that `flatten` doesn't ever copy. I also checked and the old implementation returned contiguous as well,  so this should be fine.

peterbell10 committed:
Update on "Improve repeat_interleave with scalar repeat value"


`repeat_interleave_symint` is currently implemented by guarding on the `SymInt`
and converting it to a tensor to pass to the Tensor overload. This instead
implements it as a copy of an expanded tensor, which can be done without guards
and is also much more efficient in eager mode to boot.

For example, these are timings for `x.repeat_interleave(100, dim=-1)` with `x.shape == (1000, 100)`

| Device | Time (Master) | Time (This PR)  | Speedup |
|--------|---------------|-----------------|---------|
| cpu    | 18.8 ms       | 3.5 ms          | 5.4     |
| cuda   | 271 us        | 134 us          | 2.0     |

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
